### ISSUE TITLE
ZooLock uses ZooReader which may retry on session expired

### ISSUE THREAD

**Issue by keith-turner** [apache/accumulo#1086]
Accumulo tablet servers use ephemeral nodes in zookeeper to advertise their existence. If this ephemeral node is deleted for any reason a tablet server must halt.   One way it can be deleted is if  a tablet server is not responsive enough to zookeeper servers.  In this case the zookeeper servers will close the session which causes any ephemeral nodes related to the session to be deleted.  If the tablet server is still alive and tries to communicate with zookeeper it will get a session expired exception.  When this happens the tablet server should halt.

The code that manages tablet servers ephemeral nodes is called ZookLock.  ZooLock uses an Accumulo abstraction  for Zookeeper called ZooReader.  The ZooReader code may retry on session expired events.  Need to examine ZooLock's use of ZooReader to ensure that retrying on session expired events does not delay ZooLock from becoming aware of session expiration.

**Comment by ctubbsii** [comment]
@keith-turner Any interest in investigating this before we start voting on 2.0.0 release candidates?

**Comment by keith-turner** [comment]
I added this a blocker because I think it would be good to investigate before 1.9.4.  Since its a long standing issue not introduced in 1.9.3, It need not block 1.9.4 if not investigated in time.

**Comment by keith-turner** [comment]
I looked into this a bit today.  Another issue I discovered with ZooLock's use of ZooReader is that it automatically creates new Zookeeper objects when a session has expired.  I am not sure if this is causing any problems, but its a concern I have.  The following is some background info.

 * Zookeeper objects have a one to relationship with a zookeeper session.  When a session expires zookeeper objects become non-functional.
 * Zoolock relies on ephemeral nodes which are automatically deleted when a session expires.

It may be ok that new Zookeeper objects are automatically created because the ephemeral node will cease to exist and this would be visible through the new zookeeper object.

Because of the tight coupling between zoolock, ephemeral nodes, zookeeper objects, and zookeeper sessions I think it may be better if zoolock directly uses zookeeper w/o any intermediate abstraction layer.

**Comment by milleruntime** [comment]
@keith-turner do you think this issue is still around?  I'd like to try to understand it and possibly help get it resolved for a 1.10 release.

**Comment by keith-turner** [comment]
@milleruntime I suspect its still around.  I will take a look and post some code links if I find its still an issue.

**Comment by keith-turner** [comment]
@milleruntime  I researched this a bit more, looking into the following two concerns.

 1. ZooReaderWriter retries on session expired exceptions.
 2. ZooReaderWriter automatically gets a new zookeeper when one is closed.

For issue 1, looking at the code zoolock code it deals mainly deals with session expired in watcher.  As far as I can tell ZooReaderWriter does not alter events from watcher at the moment.  For other code in ZooLock, its difficult to analyze the impact of ZooReaderWriter retrying on session expired.  There were a few places I was concerned about that seemed like they may work by accident.  So for concern 1 I have yet to identify an actual bug through code analysis, but I suspect there could be potential bugs.

For concern 2, I did find an actual race condition that seems like it could result in erroneous being placed in the metadata table.

In the code below, ZooLock calls ZooReaderWriter.getZooKeeper().getSessionId() to compute its LockId

https://github.com/apache/accumulo/blob/rel/1.9.3/fate/src/main/java/org/apache/accumulo/fate/zookeeper/ZooLock.java#L358-L363

Below is the code for ZooReaderWriter.getZooKeeper().  If you follow this code you will see that it creates a new Zookeeper if the current one is closed.  For the above code that is problematic, because a new ZK object would have a different session id.

https://github.com/apache/accumulo/blob/rel/1.9.3/fate/src/main/java/org/apache/accumulo/fate/zookeeper/ZooReaderWriter.java#L48-L54

The code below is called by the tablet server and based on the previous code may put the wrong id in the metadata table.

https://github.com/apache/accumulo/blob/rel/1.9.3/server/base/src/main/java/org/apache/accumulo/server/util/MetadataTableUtil.java#L135-L138
 
For this particular case a fix may be to make ZooLock.getLockID call ZooLock.getSessionId() instead of ZooLock.zooKeeper.getZooKeeper().getSessionId().  But I am not completely sure.  Anyway this is one example of how the zookeeper silently switching under zoolock could be problematic, there may be other cases.

**Comment by EdColeman** [comment]
Looking at the code and comparing that with the zookeeper examples on locks, would it be viable to use a guid when the ephemeral node is created?  (See https://zookeeper.apache.org/doc/r3.5.0-alpha/recipes.html#sc_recipes_errorHandlingNote - as well as the Recoverable Errors and the GUID section.)

`When creating a sequential ephemeral node there is an error case in which the create() succeeds on the server but the server crashes before returning the name of the node to the client. When the client reconnects its session is still valid and, thus, the node is not removed. The implication is that it is difficult for the client to know if its node was created or not. `

With this scheme, ZooLock may become independent of sessions - if the session that created the lock expires, then the ephemeral node goes away and without creating a new lock with a new guid there is no way to proceed.  This moves the issue from needing the session id to watching a particular node via a zk path that is unique for each time the lock is created / acquired.

Another consideration may be that getting the session id is not thread safe. The getLockId() method is synchronized, but may not want to keep repeatedly asking for the session id.

If this seems feasible, I could take a stab and see what it looks like.  Or, I just need to look more and then I would have found out why the is "a bad idea".

**Comment by dlmarion** [comment]
Created [PR](https://github.com/dlmarion/accumulo/pull/3) against the [637-validate-lock-acquired branch](https://github.com/apache/accumulo/pull/1866) branch. This new PR modifies ZooLock such that is uses the native ZooKeeper client instead of using ZooReaderWriter.
